src/cargo/util/config/mod.rs

```diff
@@ -2287,6 +2289,67 @@ impl StringList {
 #[derive(Debug, Deserialize)]
 pub struct UnmergedStringList(Vec<String>);
 
+/// A type to convert string in config values to Duration type
+///
+/// Supports deserializing duration-based string with the following format:
+/// {value}{s|ms}
+#[derive(Debug, Clone, Copy)]
+pub struct DurationString {
+    inner: Duration,
+}
+
+impl DurationString {
+    pub fn inner(&self) -> Duration {
+        self.inner
+    }
+}
+
+impl<'de> Deserialize<'de> for DurationString {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::de::Deserializer<'de>,
+    {
+        struct DurationStringVisitor;
+
+        impl<'de> serde::de::Visitor<'de> for DurationStringVisitor {
+            type Value = DurationString;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+                formatter.write_str("a value of positive integer followed by a unit (s or ms)")
+            }
+
+            fn visit_str<E>(self, s: &str) -> Result<Self::Value, E>
+            where
+                E: serde::de::Error,
+            {
+                let unit_offset = match s.find(|ch: char| ch.is_ascii_digit()) {
+                    Some(offset) => offset,
+                    None => {
+                        return Err(E::custom(format!("No unit is found on value: {}", s)));
+                    }
+                };
+                let (value_str, unit_str) = s.split_at(unit_offset);
+                let value = if let Ok(val) = value_str.parse() {
+                    val
+                } else {
+                    return Err(E::custom(format!("Invalid value format: {}", value_str)));
+                };
+                let duration = match unit_str {
+                    "s" => Duration::from_secs(value),
+                    "ms" => Duration::from_millis(value),
+                    _ => {
+                        return Err(E::unknown_variant(unit_str, &["s", "ms"]));
+                    }
+                };
+
+                Ok(DurationString { inner: duration })
+            }
+        }
+
+        deserializer.deserialize_str(DurationStringVisitor)
```

Comment:
I think this might be easier to model as :

```rust
let value: String = String::deserialize(deserializer)?;
// parse `value`...
```

Also I think this will specifically want a variety of tests.